File: 0051-N-Queens/lc0051.cc
// C++ solution.

class Solution {
public:
  vector<vector<string>> solveNQueens(int n) {
    vector<vector<string>> ans;
    vector<int> res(n, -1);
    vector<vector<bool>> states(3, vector<bool>(2*n, false));
    solve(0, states, res, ans);
    return ans;
  }

  void solve(int cur, vector<vector<bool>>& states, vector<int>& res, vector<vector<string>>& ans) {
    const int n = res.size();
    if (cur == n) {
      vector<string> checkerboard(n, string(n, '.'));
      for (int i = 0; i < n; i++){
        checkerboard[i][res[i]] = 'Q';
      }
      ans.push_back(checkerboard);
    }
    for (int i = 0; i < n; i++) {
      if (!states[0][i] && !states[1][cur+i] && !states[2][cur-i+n]) {
        res[cur] = i;
        states[0][i] = states[1][cur+i] = states[2][cur-i+n] = true;
        solve(cur+1, states, res, ans);
        states[0][i] = states[1][cur+i] = states[2][cur-i+n] = false;
      }
    }
  }
};

File: 0051-N-Queens/lc0051.rs
// Rust solution.

impl Solution {
    pub fn solve_n_queens(n: i32) -> Vec<Vec<String>> {
        let mut ans: Vec<Vec<String>> = Vec::new();
        let mut res: Vec<usize> = vec![0; n as usize];
        Self::solve(0, &mut res, &mut ans);
        ans
    }

    pub fn solve(cur: usize, res: &mut Vec<usize>, ans: &mut Vec<Vec<String>>) {
        let n = res.len();
        if (cur == n) {
            let mut checkboard: Vec<String> = Vec::new();
            for i in 0..n {
                let line = format!("{}Q{}", ".".repeat(res[i]), ".".repeat(n-res[i]-1));
                checkboard.push(line)
            }
            ans.push(checkboard);
        }
        for i in 0..n {
            let mut state: bool = true;
            for j in 0..cur {
                if i == res[j] || cur-i == j-res[j] || cur+i == j+res[j] {
                    state = false;
                    break;
                }
            }
            if state {
                res[cur] = i;
                Self::solve(cur+1, res, ans);
            }
        }
    }
}
